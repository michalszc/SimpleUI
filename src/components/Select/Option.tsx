import React, { FC, MouseEventHandler } from "react";
import styled, { css } from "styled-components";
import Styles, { StyleProps } from "../../utils/styles";
import { OptionValue } from "../../utils/types/option";

export interface OptionProps extends StyleProps {
    /**
     * Value of the option
     */
    value: OptionValue;
   /**
    * If true -> option has been selected
    */
    isChecked?: boolean;
    /**
     * If true -> option can't be selected
     */
    isDisable?: boolean;

    onClick?: MouseEventHandler<HTMLAnchorElement>;
}

const StyledOptionContainer = styled.a<OptionProps>`
    font-family: sans-serif;
    color: #555555;
    padding: 10px 5px;
    width: 100%;
    cursor: pointer;
    transition: transform .2s ease-in-out;

    &:hover {
        transform: scale(1.02);
    }

    ${({ isChecked }) => isChecked && css`
        font-size: 1.05em;
        color: #1d99ff;

    `}

`;

const Option: FC<OptionProps> = ({ onClick, value, isChecked, ...props }) => {
    return(
        <Styles {...props}>
            <StyledOptionContainer 
                value={value}
                isChecked={isChecked} 
                key={value.index}
                onClick={onClick}>
                {value.value}
            </StyledOptionContainer>
        </Styles>
    );
};

export default Option;
