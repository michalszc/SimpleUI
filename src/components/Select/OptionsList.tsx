import React, { FC } from "react";
import styled from "styled-components";
import Styles, { StyleProps } from "../../utils/styles";
import { OptionProps } from "./Option";

export interface OptionsListProps extends StyleProps {
    checkIfSelectedFunction: (v: string) => boolean;
    updateSelectedOptionsFunction: (v: string) => void;
    searchInput: string;
    children: React.ReactElement<OptionProps> | Array<React.ReactElement<OptionProps>>;
}

const OptionsList: FC<OptionsListProps> = ({ checkIfSelectedFunction, updateSelectedOptionsFunction, searchInput, children, ...props }) => {
    return(
        <Styles {...props}>
            <StyledOptionsContainer className="simpleui-select-optionslist">
                {React.Children.map(children, child => {
                        if (child.props.value.trim().toLowerCase().match(searchInput.trim().toLowerCase())) {
                            return React.cloneElement(child, {
                                isChecked: checkIfSelectedFunction(child.props.value),
                                onClick: () => updateSelectedOptionsFunction(child.props.value)
                            });
                        }
                })}
            </StyledOptionsContainer>
        </Styles>
    );
};

export default OptionsList;

const StyledOptionsContainer = styled.div`
    width: 100%;
    max-height: 200px;
    display: flex;
    flex-direction: column;
    overflow-y: auto;
    overflow-x: hidden;
    box-shadow: 0 2px 2px #cfcfcf;
    border-radius: 5px;

    &::-webkit-scrollbar {
        background-color: #dedede88;
        width: 10px;
        -webkit-overflow-scrolling: auto !important;
        background-clip: border-box;
    }

    &::-webkit-scrollbar-thumb {
        background: #a0a0a0;
        border-radius: 20px;
    }
`;
