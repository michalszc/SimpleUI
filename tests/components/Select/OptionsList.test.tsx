import React from "react";
import { render } from "@testing-library/react";
import OptionsList from "../../../src/components/Select/OptionsList";
import Option from "../../../src/components/Select/Option";

describe("OptionsList", () => {
    const data =  [
        'First option', 
        'Second option',
        'Third option',
    ];
    const options = [
        <Option value={data[0]}/>,
        <Option value={data[1]}/>,
        <Option value={data[2]}/>
    ];
    const mockFn = jest.fn();

    test("renders list with Option components", () => {
        const { container } = render(
            <OptionsList 
            searchInput="" 
            updateSelectedOptionsFunction={mockFn} 
            checkIfSelectedFunction={mockFn} 
            children={options}
            />
        );
        expect(container).toMatchSnapshot();
    });
});
