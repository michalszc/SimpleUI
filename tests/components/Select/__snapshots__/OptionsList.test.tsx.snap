// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsList renders list with Option components 1`] = `
<div>
  <div
    class="sc-bcXHqe jgetBf simpleui-select-optionslist"
  >
    <a
      class="sc-gswNZR kEHkHz"
      value="First option"
    >
      First option
    </a>
    <a
      class="sc-gswNZR kEHkHz"
      value="Second option"
    >
      Second option
    </a>
    <a
      class="sc-gswNZR kEHkHz"
      value="Third option"
    >
      Third option
    </a>
  </div>
</div>
`;
